src/objects.js

```diff
@@ -3129,36 +3127,35 @@ SpriteMorph.prototype.freshPalette = function (category) {
         ).concat(cat1.filter(
             (elem, idx) => idx % 2 === 1)
         ).concat(cat2);
-        blocks = unifiedCategories.reduce((blocks, category) =>
-            blocks.concat(
-                category === 'lists' ?
-                    [] :
-                    [
-                        this.categoryText(category),
-                        '-'
-                    ],
-                this.getPrimitiveTemplates(category),
-                '=',
-                this.customBlockTemplatesForCategory(category),
-                '='
-            ),
-            []);
+        blocks = unifiedCategories.reduce((blocks, category) => {
+            let header = [ this.categoryText(category), '-' ],
+                primitives = this.getPrimitiveTemplates(category),
+                customs = this.customBlockTemplatesForCategory(category),
+                showHeader = !['lists', 'other'].includes(category) &&
+                    (primitives.some(item => item instanceof BlockMorph) || customs.length);
+
+            return blocks.concat(
+                showHeader ? header : [],
+                primitives, '=',
+                customs, '='
+            );
+        }, []);
     } else {
         // ensure we do not modify the cached array
         blocks = this.getPrimitiveTemplates(category).slice();
     }
     blocks.push('=');
     blocks.push(this.makeBlockButton(category));
 
-    if (category === 'variables') {
-        category = ['variables', 'lists', 'other'];
```

Comment:
Well, this was my fault trying to be concise, but is the real problem, because this is what confused the hide/show primitives.